### DISCUSSION TITLE
Add Support for Paytm Statment

### DISCUSSION THREAD

**Discussion by swarajsomala** [ananthakumaran/paisa#128] [Ideas]
UPI apps are extensively used for the daily payments, can we add the import support for the UPI apps statement. Currently I have Paytm Statement and I wrote simple template to import. Due to some reasons the date format is not proper I fixed the date format manually and Imported.Please find the statement and tempalte.

```
{{#if (and (isDate ROW.J "DD-MM-YYYY") (not (isBlank ROW.F)))}}
{{ROW.J}} {{ROW.B}}
Expenses:Paytm                 {{ROW.F}} INR
Assets:Checking
{{/if}}

```
[paytm.csv](https://github.com/ananthakumaran/paisa/files/13852324/paytm.csv)

**Comment by ananthakumaran** [comment]
@swarajsomala Thanks for sharing the csv file. is this the edited date format or original date format? if it's edited could you upload with the original date format (you can change other fields though), I will see if it can be solved at the template level.

**Reply by swarajsomala** [reply]
@ananthakumaran , The uploaded file is original file I didn’t change the content. I used above template and modified statement to import, you can use the uploaded file. If you need I can provide some more input statements.

**Reply by ananthakumaran** [reply]
```handlebars
{{#if (isDate (regexpMatch ROW.A "(\d{1,2}/\d{1,2}/\d{4}).*" group=1) "D/M/YYYY")}}
  {{date (regexpMatch ROW.A "(\d{1,2}/\d{1,2}/\d{4}).*" group=1) "D/M/YYYY"}} {{ROW.B}}
  {{#if (isBlank ROW.G)}}
    {{predictAccount ROW.B prefix="Expenses"}}     {{amount ROW.F}} INR
    Assets:Checking:Paytm
  {{else}}
    Assets:Checking:Paytm                          {{amount ROW.G}} INR
    {{predictAccount ROW.B prefix="Assets:Checking"}}
  {{/if}}
{{/if}}
```

I have included the template in the repo https://github.com/ananthakumaran/paisa/commit/2014ede39a4835a290ae64c2460e9fa9172b2620

**Reply by swarajsomala** [reply]
@ananthakumaran , Thanks for adding the paytm support, Is there any way I can test it without building the paisa from scratch? Can I use the docker image which is created from that specific commit?

**Reply by ananthakumaran** [reply]
You can just copy the template above and save it. It should work with the latest version of paisa as well.

**Reply by swarajsomala** [reply]
yes will do, Thank you for the response. But In general If the commit is not yet released building is the only option?

**Reply by ananthakumaran** [reply]
yes, the docker image is also built during the release only. So the only way is to build it locally.

**Reply by swarajsomala** [reply]
ok Got it.

**Reply by swarajsomala** [reply]
Tested with the latest files it working well but now Paytm enabled a password protected statements. So For every time If I want to open the csv file I need to enter the password which is causing error with the below error.Any way to provide the password as popup?

![image](https://github.com/ananthakumaran/paisa/assets/24732012/a7e0b062-a31a-4046-9b6b-47e4ecf9a718)

**Reply by ananthakumaran** [reply]
Do you mean the XLSX file? I don't think CSV can be protected with a password. Paisa only supports password-enabled PDFs as of now, for XLSX, I haven't checked so far. Please open a new issue if this is an XLSX file, I will also try to do more research on this.

**Reply by xXWarMachineRoXx** [reply]
why are we relying on csv/xlsx?
can't we just use upi api or something, or listen for payments on a phone with a webhook etc

or maybe just use API's of existing banks ( IndusInd etc) , I do know that would require fitting for every bank provider , but its a smoother approach i think?

**Reply by ananthakumaran** [reply]
@xXWarMachineRoXx I wish it's that easy to get API access. A random app can't get anywhere near payment APIs. Only registered businesses can apply for it and it's not easy to get it even as a business.

**Reply by xXWarMachineRoXx** [reply]
@ananthakumaran 

https://sandbox.indusind.com/node/1008

I see they do have 100 calls /hour. I don't know the eligibility criterion
